File: src/main/java/ru/org/linux/user/UserModificationController.java
/*
 * Copyright 1998-2016 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Required;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.InitBinder;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import ru.org.linux.auth.AccessViolationException;
import ru.org.linux.comment.CommentService;
import ru.org.linux.comment.DeleteCommentResult;
import ru.org.linux.search.SearchQueueSender;
import ru.org.linux.site.Template;

import javax.servlet.ServletRequest;
import javax.servlet.http.HttpServletRequest;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

@Controller
public class UserModificationController {
  private static final Logger logger = LoggerFactory.getLogger(UserModificationController.class);

  private SearchQueueSender searchQueueSender;
  private UserDao userDao;
  @Autowired
  private CommentService commentService;

  @Autowired
  @Required
  public void setSearchQueueSender(SearchQueueSender searchQueueSender) {
    this.searchQueueSender = searchQueueSender;
  }

  @Autowired
  public void setUserDao(UserDao userDao) {
    this.userDao = userDao;
  }

  /**
   * Возвращает объект User модератора, если текущая сессия не модераторская, тогда исключение
   * @param request текущий http запрос
   * @return текущий модератор
   * @throws Exception если модератора нет
   */
  private static User getModerator(HttpServletRequest request) throws Exception {
    Template tmpl = Template.getTemplate(request);
    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not moderator");
    }
    return tmpl.getCurrentUser();
  }

  /**
   * Контроллер блокировки пользователя
   * @param request http запрос
   * @param user блокируемый пользователь
   * @param reason причина блокировки
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или блокируемого пользователя
   * нельзя блокировать
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=block")
  public ModelAndView blockUser(
      HttpServletRequest request,
      @RequestParam("id") User user,
      @RequestParam(value = "reason", required = false) String reason
  ) throws Exception {
    User moderator = getModerator(request);
    if (!user.isBlockable() && !moderator.isAdministrator()) {
      throw new AccessViolationException("Пользователя " + user.getNick() + " нельзя заблокировать");
    }

    if (user.isBlocked()) {
      throw new UserErrorException("Пользователь уже блокирован");
    }

    userDao.block(user, moderator, reason);
    logger.info("User " + user.getNick() + " blocked by " + moderator.getNick());
    return redirectToProfile(user);
  }

  /**
   * Выставляем score=50 для пользователей у которых score меньше
   *
   * @param request http запрос
   * @param user кому ставим score
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или пользователь блокирован
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=score50")
  public ModelAndView score50(
          HttpServletRequest request,
          @RequestParam("id") User user
  ) throws Exception {
    User moderator = getModerator(request);
    if (user.isBlocked() || user.isAnonymous()) {
      throw new AccessViolationException("Нельзя выставить score=50 пользователю " + user.getNick());
    }

    userDao.score50(user, moderator);

    return redirectToProfile(user);
  }

  /**
   * Контроллер разблокировки пользователя
   * @param request http запрос
   * @param user разблокируемый пользователь
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или пользователя нельзя разблокировать
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=unblock")
  public ModelAndView unblockUser(
      HttpServletRequest request,
      @RequestParam("id") User user
  ) throws Exception {

    User moderator = getModerator(request);
    if (!user.isBlockable() && !moderator.isAdministrator()) {
      throw new AccessViolationException("Пользователя " + user.getNick() + " нельзя разблокировать");
    }
    userDao.unblock(user, moderator);
    logger.info("User " + user.getNick() + " unblocked by " + moderator.getNick());
    return redirectToProfile(user);
  }

  private static ModelAndView redirectToProfile(User user) throws UnsupportedEncodingException{
    return new ModelAndView(new RedirectView(getNoCacheLinkToProfile(user)));
  }

  private static String getNoCacheLinkToProfile(User user) throws UnsupportedEncodingException{
    Random random = new Random();
    return "/people/" + URLEncoder.encode(user.getNick(), "UTF-8") + "/profile?nocache=" + random.nextInt();
  }

  /**
   * Контроллер блокирования и полного удаления комментариев и топиков пользователя
   * @param request http запрос
   * @param user блокируемый пользователь
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или пользователя нельзя блокировать
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=block-n-delete-comments")
  public ModelAndView blockAndMassiveDeleteCommentUser(
      HttpServletRequest request,
      @RequestParam("id") User user,
      @RequestParam(value = "reason", required = false) String reason
  ) throws Exception {
    User moderator = getModerator(request);
    if (!user.isBlockable() && !moderator.isAdministrator()) {
      throw new AccessViolationException("Пользователя " + user.getNick() + " нельзя заблокировать");
    }

    if (user.isBlocked()) {
      throw new UserErrorException("Пользователь уже блокирован");
    }

    Map<String, Object> params = new HashMap<>();
    params.put("message", "Удалено");
    DeleteCommentResult deleteCommentResult = commentService.deleteAllCommentsAndBlock(user, moderator, reason);

    logger.info("User " + user.getNick() + " blocked by " + moderator.getNick());

    params.put("bigMessage",
            "Удалено комментариев: "+deleteCommentResult.getDeletedCommentIds().size()+"<br>"+
            "Удалено тем: "+deleteCommentResult.getDeletedTopicIds().size()
    );

    for (int topicId : deleteCommentResult.getDeletedTopicIds()) {
      searchQueueSender.updateMessage(topicId, true);
    }

    searchQueueSender.updateComment(deleteCommentResult.getDeletedCommentIds());

    return new ModelAndView("action-done", params);
  }

  /**
   * Контроллер смена признака корректора
   * @param request http запрос
   * @param user блокируемый пользователь
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или пользователя нельзя сделать корректором
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=toggle_corrector")
  public ModelAndView toggleUserCorrector(
      HttpServletRequest request,
      @RequestParam("id") User user
  ) throws Exception {
    User moderator = getModerator(request);
    if (user.getScore()<User.CORRECTOR_SCORE) {
      throw new AccessViolationException("Пользователя " + user.getNick() + " нельзя сделать корректором");
    }
    userDao.toggleCorrector(user, moderator);
    logger.info("Toggle corrector " + user.getNick() + " by " + moderator.getNick());

    return redirectToProfile(user);
  }

  /**
   * Сброс пароля пользователю
   * @param request http запрос
   * @param user пользователь которому сбрасываем пароль
   * @return сообщение о успешности сброса
   * @throws Exception при ошибке или отсутствии прав
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=reset-password")
  public ModelAndView resetPassword(
      HttpServletRequest request,
      @RequestParam("id") User user
  ) throws Exception {
    User moderator = getModerator(request);

    if (user.isModerator() || user.isAnonymous()) {
      throw new AccessViolationException("Пользователю " + user.getNick() + " нельзя сбросить пароль");
    }

    userDao.resetPassword(user, moderator);

    logger.info("Пароль "+user.getNick()+" сброшен модератором "+moderator.getNick());

    ModelAndView mv = new ModelAndView("action-done");
    mv.getModel().put("link", getNoCacheLinkToProfile(user));
    mv.getModel().put("message", "Пароль сброшен");
    return mv;
  }
  
  /**
   * Контроллер отчистки дополнительной информации в профиле
   * @param request http запрос
   * @param user блокируемый пользователь
   * @return возвращаемся в профиль
   * @throws Exception обычно если текущий пользователь не модератор или нельзя трогать дополнительные сведения
   */
  @RequestMapping(value = "/usermod.jsp", method = RequestMethod.POST, params = "action=remove_userinfo")
  public ModelAndView removeUserInfo(
      HttpServletRequest request,
      @RequestParam("id") User user
  ) throws Exception {
    User moderator = getModerator(request);
    if (user.isModerator()) {
      throw new AccessViolationException("Пользователю " + user.getNick() + " нельзя удалить сведения");
    }
    userDao.removeUserInfo(user, moderator);
    logger.info("Clearing " + user.getNick() + " userinfo by " + moderator.getNick());

    return redirectToProfile(user);
  }

  @RequestMapping(value="/remove-userpic.jsp", method= RequestMethod.POST)
  public ModelAndView removeUserpic(
    ServletRequest request,
    @RequestParam("id") User user
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isSessionAuthorized()) {
      throw new AccessViolationException("Not autorized");
    }

    User currentUser = tmpl.getCurrentUser();

    // Не модератор не может удалять чужие аватары
    if (!tmpl.isModeratorSession() && currentUser.getId()!=user.getId()) {
      throw new AccessViolationException("Not permitted");
    }

    if (userDao.resetUserpic(user, currentUser)) {
      logger.info("Clearing " + user.getNick() + " userpic by " + currentUser.getNick());
    } else {
      logger.debug("SKIP Clearing " + user.getNick() + " userpic by " + currentUser.getNick());
    }

    return redirectToProfile(user);
  }

  @InitBinder
  public void initBinder(WebDataBinder binder) {
    binder.registerCustomEditor(User.class, new UserIdPropertyEditor(userDao));
  }
}

File: src/main/scala/ru/org/linux/user/UserModificationController.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.user

import com.typesafe.scalalogging.StrictLogging
import org.springframework.stereotype.Controller
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import ru.org.linux.auth.AccessViolationException
import ru.org.linux.auth.AuthUtil.{AuthorizedOnly, ModeratorOnly}
import ru.org.linux.comment.CommentDeleteService
import ru.org.linux.search.SearchQueueSender

import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.sql.Timestamp
import java.time.Duration
import java.util.concurrent.ThreadLocalRandom
import scala.collection.mutable
import scala.jdk.CollectionConverters.{ListHasAsScala, MapHasAsJava}

@Controller
object UserModificationController {
  private def redirectToProfile(user: User) = new ModelAndView(new RedirectView(getNoCacheLinkToProfile(user)))

  private def getNoCacheLinkToProfile(user: User) = {
    "/people/" + URLEncoder.encode(user.getNick, StandardCharsets.UTF_8) + "/profile?nocache=" + ThreadLocalRandom.current().nextInt
  }

  // get 'now', add the duration and returns result;
  // the duration can be negative
  private def getUntil(shift: String): Timestamp = {
    val d = Duration.parse(shift)
    val now = new Timestamp(System.currentTimeMillis)
    now.setTime(now.getTime + d.toMillis)
    now
  }
}

@Controller
class UserModificationController(searchQueueSender: SearchQueueSender, userDao: UserDao,
                                 commentService: CommentDeleteService, userService: UserService) extends StrictLogging {
  import UserModificationController.*

  /**
   * Контроллер блокировки пользователя
   *
   * @param user   блокируемый пользователь
   * @param reason причина блокировки
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=block"))
  def blockUser(@RequestParam("id") user: User,
                @RequestParam(value = "reason", required = false) reason: String): ModelAndView = ModeratorOnly { moderator =>
    if (!userService.isBlockable(user = user, by = moderator.user)) {
      throw new AccessViolationException(s"Пользователя ${user.getNick} нельзя заблокировать")
    }

    if (user.isBlocked) {
      throw new UserErrorException("Пользователь уже блокирован")
    }

    userDao.block(user, moderator.user, reason)

    logger.info(s"User ${user.getNick} blocked by ${moderator.user.getNick}")

    redirectToProfile(user)
  }

  /**
   * Выставляем score=50 для пользователей у которых score меньше
   *
   * @param user кому ставим score
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=score50"))
  def score50(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.isBlocked || user.isAnonymous || user.getScore > 50) {
      throw new AccessViolationException(s"Нельзя выставить score=50 пользователю ${user.getNick}")
    }

    userDao.score50(user, moderator.user)

    redirectToProfile(user)
  }

  /**
   * Контроллер разблокировки пользователя
   *
   * @param user разблокируемый пользователь
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=unblock"))
  def unblockUser(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (!userService.isBlockable(user = user, by = moderator.user)) {
      throw new AccessViolationException(s"Пользователя ${user.getNick} нельзя разблокировать")
    }

    userDao.unblock(user, moderator.user)

    logger.info(s"User ${user.getNick} unblocked by ${moderator.user.getNick}")

    redirectToProfile(user)
  }

  /**
   * Контроллер блокирования и полного удаления комментариев и топиков пользователя
   *
   * @param user блокируемый пользователь
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=block-n-delete-comments"))
  def blockAndMassiveDeleteCommentUser(@RequestParam("id") user: User,
                                       @RequestParam(value = "reason", required = false) reason: String): ModelAndView = ModeratorOnly { moderator =>
    if (!userService.isBlockable(user = user, by = moderator.user)) {
      throw new AccessViolationException(s"Пользователя ${user.getNick} нельзя заблокировать")
    }

    if (user.isBlocked) {
      throw new UserErrorException("Пользователь уже блокирован")
    }

    val params = new mutable.HashMap[String, AnyRef]
    params.put("message", "Удалено")
    val deleteCommentResult = commentService.deleteAllCommentsAndBlock(user, moderator.user, reason)

    logger.info(s"User ${user.getNick} blocked by ${moderator.user.getNick}")

    params.put("bigMessage", s"Удалено комментариев: ${deleteCommentResult.getDeletedCommentIds.size}<br>Удалено тем: ${deleteCommentResult.getDeletedTopicIds.size}")

    deleteCommentResult.getDeletedTopicIds.asScala.foreach { topicId =>
      searchQueueSender.updateMessage(topicId, true)
    }

    searchQueueSender.updateComment(deleteCommentResult.getDeletedCommentIds)

    new ModelAndView("action-done", params.asJava)
  }

  /**
   * Контроллер смена признака корректора
   *
   * @param user блокируемый пользователь
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=toggle_corrector"))
  def toggleUserCorrector(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.getScore < UserService.CorrectorScore) {
      throw new AccessViolationException(s"Пользователя ${user.getNick} нельзя сделать корректором")
    }

    userDao.toggleCorrector(user, moderator.user)

    logger.info(s"Toggle corrector ${user.getNick} by ${moderator.user.getNick}")

    redirectToProfile(user)
  }

  /**
   * Сброс пароля пользователю
   *
   * @param user пользователь которому сбрасываем пароль
   * @return сообщение о успешности сброса
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=reset-password"))
  def resetPassword(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.isModerator && !moderator.user.isAdministrator) {
      throw new AccessViolationException(s"Пользователю ${user.getNick} нельзя сбросить пароль")
    }

    if (user.isAnonymous) {
      throw new AccessViolationException(s"Пользователю ${user.getNick} нельзя сбросить пароль")
    }

    userDao.resetPassword(user, moderator.user)

    logger.info(s"Пароль ${user.getNick} сброшен модератором ${moderator.user.getNick}")

    val mv = new ModelAndView("action-done")

    mv.getModel.put("link", getNoCacheLinkToProfile(user))
    mv.getModel.put("message", "Пароль сброшен")

    mv
  }

  /**
   * Контроллер отчистки дополнительной информации в профиле
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=remove_userinfo"))
  def removeUserInfo(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.isAnonymous) {
      throw new AccessViolationException(s"Пользователю ${user.getNick} нельзя удалить сведения")
    }

    userService.removeUserInfo(user, moderator.user)

    UserModificationController.redirectToProfile(user)
  }

  /**
   * Контроллер отчистки поля город
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=remove_town"))
  def removeTown(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.isAnonymous) {
      throw new AccessViolationException(s"Пользователю ${user.getNick} нельзя удалить сведения")
    }

    userService.removeTown(user, moderator.user)

    redirectToProfile(user)
  }

  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=remove_url"))
  def removeUrl(@RequestParam("id") user: User): ModelAndView = ModeratorOnly { moderator =>
    if (user.isAnonymous) {
      throw new AccessViolationException(s"Пользователю ${user.getNick} нельзя удалить сведения")
    }

    userService.removeUrl(user, moderator.user)

    redirectToProfile(user)
  }

  @RequestMapping(value = Array("/remove-userpic.jsp"), method = Array(RequestMethod.POST))
  def removeUserpic(@RequestParam("id") user: User): ModelAndView = AuthorizedOnly { currentUser =>
    // Не модератор не может удалять чужие аватары
    if (!currentUser.moderator && currentUser.user.getId != user.getId) {
      throw new AccessViolationException("Not permitted")
    }

    if (userDao.resetUserpic(user, currentUser.user)) {
      logger.info(s"Clearing ${user.getNick} userpic by ${currentUser.user.getNick}")
    }

    redirectToProfile(user)
  }

  /**
   * Контроллер заморозки и разморозки пользователя
   *
   * @param user   блокируемый пользователь
   * @param reason причина заморозки, общедоступна в дальнейшем
   * @param shift  отсчёт времени от текущей точки, может быть отрицательным, в
   *               в результате даёт until, отрицательное значение используется
   *               для разморозки
   * @return возвращаемся в профиль
   */
  @RequestMapping(value = Array("/usermod.jsp"), method = Array(RequestMethod.POST), params = Array("action=freeze"))
  def freezeUser(@RequestParam(name = "id") user: User, @RequestParam(name = "reason") reason: String,
                 @RequestParam(name = "shift") shift: String): ModelAndView = ModeratorOnly { moderator =>
    if (reason.length > 255) {
      throw new UserErrorException("Причина слишком длиная, максимум 255 байт")
    }

    val until = UserModificationController.getUntil(shift)

    if (!userService.isFreezable(user = user, by = moderator.user)) {
      throw new AccessViolationException(s"Пользователя ${user.getNick} нельзя заморозить")
    }

    if (user.isBlocked) {
      throw new UserErrorException("Пользователь блокирован, его нельзя заморозить")
    }

    userDao.freezeUser(user, moderator.user, reason, until)

    logger.info(s"Freeze ${user.getNick} by ${moderator.user.getNick} until $until")

    redirectToProfile(user)
  }

  @RequestMapping(value = Array("/people/{nick}/profile/wipe"), method = Array(RequestMethod.GET, RequestMethod.HEAD))
  def wipe(@PathVariable nick: String): ModelAndView = ModeratorOnly { moderator =>
    val user = userService.getUser(nick)

    if (!userService.isBlockable(user = user, by = moderator.user)) {
      throw new AccessViolationException("Пользователя нельзя заблокировать")
    }

    if (user.isBlocked) {
      throw new UserErrorException("Пользователь уже блокирован")
    }

    val mv = new ModelAndView("wipe-user")

    mv.getModel.put("user", user)
    mv.getModel.put("commentCount", userDao.getExactCommentCount(user))

    mv
  }

  @InitBinder
  def initBinder(binder: WebDataBinder): Unit = {
    binder.registerCustomEditor(classOf[User], new UserIdPropertyEditor(userService))
  }
}